### PR TITLE
Uplift elaboration into `rustc_type_ir`

### PR DESCRIPTION
**Pull request by compiler-errors**
Allows us to deduplicate and consolidate elaboration (including these stupid elaboration duplicate fns i added for pretty printing like 3 years ago) so I'm pretty hyped about this change :3

r? lcnr

### CODE REVIEW

File: compiler/rustc_type_ir/src/interner.rs [hunk 1]
```diff
@@ -213,7 +213,12 @@ pub trait Interner:
     fn explicit_super_predicates_of(
         self,
         def_id: Self::DefId,
-    ) -> ty::EarlyBinder<Self, impl IntoIterator<Item = Self::Clause>>;
+    ) -> ty::EarlyBinder<Self, impl IntoIterator<Item = (Self::Clause, Self::Span)>>;
```
Comment (compiler-errors) [inline compiler/rustc_type_ir/src/interner.rs:216]: This is a somewhat dissatisfying incongruency between these fns and all the others above, but we need these in order to give good spans when elaborating 😆. 🤷

File: compiler/rustc_type_ir/src/interner.rs [hunk 2]
```diff
@@ -268,6 +270,11 @@ pub trait Interner:
         param_env: Self::ParamEnv,
         placeholder: Self::PlaceholderConst,
     ) -> Self::Ty;
+
+    fn anonymize_bound_vars<T: TypeFoldable<Self>>(
```
Comment (compiler-errors) [inline compiler/rustc_type_ir/src/interner.rs:274]: I'll uplift this later.

File: compiler/rustc_middle/src/ty/context.rs [hunk 3]
```diff
@@ -347,12 +347,18 @@ impl<'tcx> Interner for TyCtxt<'tcx> {
     fn explicit_super_predicates_of(
         self,
         def_id: DefId,
-    ) -> ty::EarlyBinder<'tcx, impl IntoIterator<Item = ty::Clause<'tcx>>> {
+    ) -> ty::EarlyBinder<'tcx, impl IntoIterator<Item = (ty::Clause<'tcx>, Span)>> {
+        ty::EarlyBinder::bind(
+            self.explicit_super_predicates_of(def_id).instantiate_identity(self).into_iter(),
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/context.rs:352]: that `into_iter` isn't necessary given that the return type is `IntoIterator`?

File: compiler/rustc_middle/src/ty/context.rs [hunk 4]
```diff
@@ -347,12 +347,18 @@ impl<'tcx> Interner for TyCtxt<'tcx> {
...
+    fn explicit_implied_predicates_of(
+        self,
+        def_id: DefId,
+    ) -> ty::EarlyBinder<'tcx, impl IntoIterator<Item = (ty::Clause<'tcx>, Span)>> {
         ty::EarlyBinder::bind(
-            self.explicit_super_predicates_of(def_id)
-                .instantiate_identity(self)
-                .predicates
-                .into_iter(),
+            self.explicit_implied_predicates_of(def_id).instantiate_identity(self).into_iter(),
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/context.rs:361]: same here?

File: compiler/rustc_type_ir/src/elaborate.rs [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    elaborator
+}
+
+impl<I: Interner, O: Elaboratable<I>> Elaborator<I, O> {
+    fn extend_deduped(&mut self, obligations: impl IntoIterator<Item = O>) {
+        // Only keep those bounds that we haven't already seen.
+        // This is necessary to prevent infinite recursion in some
+        // cases. One common case is when people define
+        // `trait Sized: Sized { }` rather than `trait Sized { }`.
+        // let visited = &mut self.visited;
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/elaborate.rs:62]:
```suggestion
```

File: compiler/rustc_type_ir/src/elaborate.rs [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+                                    r_min,
+                                )))
+                            }
+
+                            Component::EscapingAlias(_) => {
+                                // We might be able to do more here, but we don't
+                                // want to deal with escaping vars right now.
+                                None
+                            }
+                        })
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/elaborate.rs:180]: maybe move this into `fn component_to_clause` 🤔

**Comment by lcnr** [comment]
nice, some nits and suggestions, otherwise r=me

**Comment by compiler-errors** [comment]
@rust-timer build 194c829aa44322a8ee37f88a0a46d822f1c4d881

**Pull request merged**